signer/xpi/cose.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	for i, sig := range msg.Signatures {
+		eeCert, sigErr := isValidCOSESignature(&sig)
+		if sigErr != nil {
+			err = errors.Wrapf(sigErr, "xpi: cose signature %d is invalid", i)
+			return
+		}
+		eeCerts = append(eeCerts, eeCert)
+	}
+
+	return
+}
+
+// verifyCOSESignatures checks that:
+//
+// 1) COSE manifest and signature files are present
+// 2) the PKCS7 manifest is present
+// 3) the COSE and PKCS7 manifests do not include COSE files
+// 4) we can decode the COSE signature and it has the right format for an XPI
+// 5) the right number of signatures are present and all intermediate and end entity certs parse properly
+// TODO: 6) there is a trusted path from the included COSE EE certs to the signer cert using the provided intermediates
+//
+func verifyCOSESignatures(signedFile signer.SignedFile, truststore *x509.CertPool, signOptions Options) error {
+	coseManifest, err := readFileFromZIP(signedFile, "META-INF/cose.manifest")
+	if err != nil {
+		return fmt.Errorf("xpi: failed to read META-INF/cose.manifest from signed zip: %v", err)
+	}
+	coseMsgBytes, err := readFileFromZIP(signedFile, "META-INF/cose.sig")
+	if err != nil {
+		return fmt.Errorf("xpi: failed to read META-INF/cose.sig from signed zip: %v", err)
+	}
+	pkcs7Manifest, err := readFileFromZIP(signedFile, "META-INF/manifest.mf")
+	if err != nil {
+		return fmt.Errorf("xpi: failed to read META-INF/manifest.mf from signed zip: %v", err)
+	}
+
+	var coseFileNames = [][]byte{
+		[]byte("Name: META-INF/cose.sig"),
+		[]byte("Name: META-INF/cose.manifest"),
+	}
+	for _, coseFileName := range coseFileNames {
+		if !bytes.Contains(pkcs7Manifest, coseFileName) {
+			return fmt.Errorf("xpi: pkcs7 manifest does not contain the line: %s", coseFileName)
+		}
+
+		if bytes.Contains(coseManifest, coseFileName) {
+			return fmt.Errorf("xpi: cose manifest contains the line: %s", coseFileName)
+		}
+	}
+
+	xpiSig, unmarshalErr := Unmarshal(base64.StdEncoding.EncodeToString(coseMsgBytes), nil)
+	if unmarshalErr != nil {
+		return errors.Wrap(unmarshalErr, "xpi: error unmarshaling cose.sig")
+	}
+	if xpiSig != nil && xpiSig.signMessage != nil && len(xpiSig.signMessage.Signatures) != len(signOptions.COSEAlgorithms) {
+		return fmt.Errorf("xpi: cose.sig contains %d signatures, but expected %d", len(xpiSig.signMessage.Signatures), len(signOptions.COSEAlgorithms))
+	}
+
+	intermediateCerts, eeCerts, err := isValidCOSEMessage(xpiSig.signMessage)
+	if err != nil {
+		return errors.Wrap(err, "xpi: cose.sig is not a valid COSE SignMessage")
+	}
+
+	// check that we can verify EE certs with the provided intermediates
+	intermediates := x509.NewCertPool()
+	for _, intermediateCert := range intermediateCerts {
+		intermediates.AddCert(intermediateCert)
+	}
+	cndigest := sha256.Sum256([]byte(signOptions.ID))
+	dnsName := fmt.Sprintf("%x.%x.addons.mozilla.org", cndigest[:16], cndigest[16:])
+
+	for i, eeCert := range eeCerts {
+		opts := x509.VerifyOptions{
```

Comment:
you should also verify that the signOptions.ID matches the subject.CommonName of the EE